File: Backend/Institute_Management_System/src/main/java/com/ims/service/TeacherService.java
package com.ims.service;

// ===================== IMPORTS =====================

// Java collection utilities
import java.util.List;

// DTO imports used by TeacherService
import com.ims.dto.MarksSubmitRequestDTO;
import com.ims.dto.MySubjectDTO;
import com.ims.dto.NoticeResponse;
import com.ims.dto.StudentAttendanceDTO;
import com.ims.dto.StudentListDTO;
import com.ims.dto.StudentMarksViewDTO;
import com.ims.dto.TeacherMyAttendanceDTO;
import com.ims.dto.TeacherProfileResponse;
import com.ims.dto.TeacherProfileUpdateRequest;
import com.ims.dto.TeacherScheduleDTO;

// Entity import (currently not used directly, but part of domain)
import com.ims.entity.Role;

// ===================== SERVICE INTERFACE =====================

/**
 * TeacherService
 * 
 * This interface defines all business operations related to
 * teacher functionality such as dashboard data, profile,
 * students, attendance, subjects, exams, and notices.
 */
public interface TeacherService {

    /* ================= Teacher DASHBOARD ================= */

    // Count total students assigned to a specific teacher
    long countStudentsForTeacher(Long teacherId);

    // Count total subjects assigned to a teacher
    long countAssignedSubjects(Long teacherId);

    // Get average monthly attendance in present/total format
    String getAverageMonthlyAttendance();

    // Get latest notices for teacher dashboard
    List<NoticeResponse> getLatestNoticesForTeacher();

    

    /* ================= TEACHER PROFILE ================= */

    // Fetch teacher profile details
    TeacherProfileResponse getTeacherProfile(Long userId);

    // Update editable teacher profile fields
    TeacherProfileResponse updateTeacherProfile(
            Long userId,
            TeacherProfileUpdateRequest request
    );

    
    
    
    /* ================= STUDENTS Information ================= */

    // Get list of all students
    List<StudentListDTO> getAllStudents();

    // Get students for marks entry using course and subject
    List<StudentMarksViewDTO> getStudentsForMarksByName(
            String courseName,
            String subjectName
    );
    
    // Get subjects assigned to a teacher
    List<MySubjectDTO> getMySubjects(Long teacherId);

    // Submit or update marks for students
    void submitMarks(MarksSubmitRequestDTO request);

    // Get attendance list of students (teacher view)
    List<StudentAttendanceDTO> getStudentAttendanceList();
    
    
    

    /* ================= Teacher Subjects ========================= */

    // Get teacher schedule (classes, subjects, timings)
    List<TeacherScheduleDTO> getTeacherSchedule(Long teacherId);

    

    
    
    
    /* ================= Teacher Attendance ================= */

    // Get logged-in teacher's own attendance records
    List<TeacherMyAttendanceDTO> getMyAttendance(Long teacherId);

    /* ================= Teacher Exam Dashboard ================= */
    
    

    /* Exam-related operations are handled in ExamServiceImpl */
    
    
    
    /* ================= Notices ================= */

    // Get all notices visible to teachers
    List<NoticeResponse> getAllNoticesForTeacher();

}

File: frontend/app/src/services/teacherService.js
import API from "../api";

const BASE_URL = "/teacher";

// 👉 Get total active students count
export const getStudentCount = () => {
    return API.get(`${BASE_URL}`);
};

export const getStudentCountForTeacher = (teacherId) => {
    return API.get(`${BASE_URL}/student-count/${teacherId}`);
};

/////
export const getAverageAttendance = () => {
    return API.get(`${BASE_URL}/average-attendance`);
};




// GET /teacher/allstudents
export const getAllStudents = () => {
    return API.get(`${BASE_URL}/allstudents`);
};

// ================= Evaluate Student =================
export const getStudentsForMarksByName = (courseName, subjectName) => {
    return API.get(`${BASE_URL}/students-for-marks/by-name`, {
        params: { courseName, subjectName },
    });
};

export const submitMarks = (payload) =>
    API.post(`${BASE_URL}/submit-marks`, payload);


// ================================================

// 🔹 Get student attendance list
export const getStudentAttendance = () => {
    return API.get(`${BASE_URL}/attendance`);
};

// Subjects (FIXED)
export const getTeacherSubjects = (teacherId) =>
    API.get(`${BASE_URL}/subjects/teacher/${teacherId}`);


/* ================= PROFILE ================= */
export const fetchTeacherProfile = (teacherId) =>
    API.get(`${BASE_URL}/profile/${teacherId}`);

export const updateTeacherProfile = (teacherId, payload) =>
    API.put(`${BASE_URL}/update/${teacherId}`, payload);

/* ================= TEACHER NOTICES ================= */

export const getTeacherNotices = () => {
    return API.get(`${BASE_URL}/notices`);
};

/**
 * 🔹 Get TOP / RECENT notices for teacher (Dashboard)
 * API: GET /teacher/notices/top
 * Response: { data: [...], status: "SUCCESS" }
 */
export const getTopTeacherNotices = () => {
    return API.get(`${BASE_URL}/notices/top`);
};
// ================= TEACHER SCHEDULE =================
export const getTeacherSchedule = (teacherId) => {
    return API.get(`${BASE_URL}/schedule/${teacherId}`);
};

export const getAssignedSubjectCount = (teacherId) => {
    return API.get(`${BASE_URL}/subject-count/${teacherId}`);
};


// GET /api/teacher/my-attendance/{teacherId}
export const getTeacherMyAttendance = (teacherId) => {
  return API.get(`/teacher/my-attendance/${teacherId}`);
};